File: examples/ts/reading_generator.ts
/**
 * Reading (furigana) generator - generate HTML ruby annotations
 *
 * Demonstrates: morpheme-level analysis, reading/conjugation fields,
 * and generating furigana HTML for web display.
 *
 * Run: npx tsx examples/ts/reading_generator.ts
 */
import { Suzume } from '../../dist/index.js';

const suzume = await Suzume.create();

const text = '彼女は図書館で難しい本を読んでいた';
const morphemes = suzume.analyze(text);

// Table output
console.log('Surface\tReading\tPOS\tBase\tConj');
console.log('------\t-------\t---\t----\t----');
for (const m of morphemes) {
  console.log(
    `${m.surface}\t${m.reading}\t${m.pos}\t${m.baseForm}\t${m.conjForm ?? '-'}`,
  );
}

// Generate furigana HTML
function toRubyHtml(morphemes: ReturnType<typeof suzume.analyze>): string {
  return morphemes
    .map((m) => {
      if (!m.reading || m.surface === m.reading) return m.surface;
      return `<ruby>${m.surface}<rp>(</rp><rt>${m.reading}</rt><rp>)</rp></ruby>`;
    })
    .join('');
}

console.log('\nFurigana HTML:');
console.log(toRubyHtml(morphemes));

File: examples/cpp/reading_generator.cpp
/**
 * Reading (furigana) generator - extract readings for kanji words
 *
 * Demonstrates: morpheme-level analysis, reading field,
 * conjugation info, and lemma access.
 */
#include <iostream>
#include <string_view>

#include "core/types.h"
#include "grammar/conjugation.h"
#include "suzume.h"

int main() {
  suzume::Suzume analyzer;

  std::string_view text = "彼女は図書館で難しい本を読んでいた";

  auto morphemes = analyzer.analyze(text);

  // Display with readings
  std::cout << "Surface\tReading\tPOS\tLemma\tConj\n";
  std::cout << "------\t-------\t---\t-----\t----\n";

  for (const auto& m : morphemes) {
    std::cout << m.surface << "\t" << m.reading << "\t"
              << suzume::core::posToString(m.pos) << "\t"
              << m.getLemma() << "\t";

    if (m.conj_form != suzume::grammar::ConjForm::Base) {
      std::cout << suzume::grammar::conjFormToString(m.conj_form);
    } else {
      std::cout << "-";
    }
    std::cout << "\n";
  }

  // Generate furigana HTML
  std::cout << "\nFurigana HTML:\n";
  for (const auto& m : morphemes) {
    if (m.reading.empty() || m.surface == m.reading) {
      std::cout << m.surface;
    } else {
      std::cout << "<ruby>" << m.surface << "<rp>(</rp><rt>" << m.reading
                << "</rt><rp>)</rp></ruby>";
    }
  }
  std::cout << "\n";

  return 0;
}
